### ISSUE TITLE
ENH: Allow loadtxt to ignore a single column

### ISSUE THREAD

**Issue by jakobjakobson13** [numpy/numpy#13878]
In the loadtxt routine you have to explicitly specify all the columns you want to read in as soon as you want to discard one column. However, I normally use the first row and first column for labels. So my data looks similar to the following example:
```python
>>> import io
>>> import numpy as np
>>> a = " \tsample1\tsample2\n280K\t0\t1\n300K\t2\t3"
>>> print(a)
 	sample1	sample2
280K	0	1
300K	2	3
```

To import the data to numy the only way that worked for me was the following command:
```python
np.loadtxt(io.StringIO(a), skiprows=1, usecols=(1, 2))
```

Unfortunately this becomes very unhandy for longer tables. So could you provide a solution similar to `skiprows` or an inverted selection of columns in a way like `usecols=(1:)` or something that way?

Regards

**Comment by WarrenWeckesser** [comment]
Over in https://github.com/numpy/numpy/pull/15080, @daniel-haines has added an argument to `loadtxt` that allows you to skip the first `n` columns.  Before getting into a review of that pull request, we should think about other possible APIs for this.

The problem described above is that currently there is no easy way to tell `loadtxt` to skip, say, the first 3 columns and return the rest, unless you already know how many columns are in the file.

Another selection not possible with the current API is to read, say, the *last two* columns, regardless of how many columns are actually in the file.

I think all the Python programmers are now thinking "Ah, what you want is a *slice* of the columns."  E.g. if the list `cols = list(range(10))` represents the columns, we would write `cols[3:]` to skip the first three columns, and `cols[-2:]` get the last two columns.

Python syntax does not allow us to write `usecols=3:` or `usecols=-2:`, but we could use the built-in [`slice`](https://docs.python.org/3/library/functions.html?highlight=slice#slice) object to represent the desired columns.  We would write `usecols=slice(3, None)` to skip the first three columns, `usecols=slice(-2, None)` to read just the last two columns, etc.

What do folks think of that API?

**Comment by eric-wieser** [comment]
Allowing a lambda might be an option too, `skipcols=lambda n: set(range(n)) - {0, n-1}` would skip all but the first and last column

**Comment by eric-wieser** [comment]
> Python syntax does not allow us to write `usecols=3:` or `usecols=-2:`, but we could use the built-in [`slice`](https://docs.python.org/3/library/functions.html?highlight=slice#slice) object to represent the desired columns. We would write `usecols=slice(3, None)` to skip the first three columns, `usecols=slice(-2, None)` to read just the last two columns, etc.

Don't forget we have `np.s_[3:]` and `np.s_[-2:]` to make that spelling more natural.

**Comment by WarrenWeckesser** [comment]
@eric-wieser, thanks, I *did* forget!

A brief review...

Currently, `usecols` must be `List[int]` (or just `int`).  The values can be negative to select from the right end.  (Also, the order is signficant.  The columns can be rearranged, e.g. `usecols=[2, 3, 0, 1]`.)  If one does not know in advance how many columns there are, the current implementation of `usecols` does not allow one to, for example, skip just the first column and read the rest.

I suggested above allowing `usecols` to accept a `slice` object.  Then you can do things like:
* `usecols=np.s_[1:]` to skip the first column,
* `usecols=np.s_[1:-1]` to skip first and last column,
* `usecols=np.s_[-3:]` to read just the last three columns
* etc.

This would cover quite a few use-cases, but it is limited to what can be expressed as a slice.  Suppose you want to skip just the second column?  That set can't be represented as a slice, and it can't be represented as a list of column indices when you don't know in advance how many columns there are.  With a `skipcols` parameter that accepts a sequence of column indices to be *skipped*, one would use `skipcols=[1]`.

@eric-wieser's suggestion to accept a callable `f(n: int)` that returns a list (or some other iterable) is probably the most general API.  It covers all possible use-cases.  To skip just the second column, you could use

      usecols=lambda n: [0] + list(range(2, n))

That's not as concise as `skipcols=[1]`, but it avoids the potential confusion about how `usecols` and `skipcols` interact, and the callable allows for any possible selection of columns.  Here are a couple examples that can't be done with the existing `usecols`, nor with a slice, nor with the proposed `skipcols`:

* Read the first half of the columns:
      `usecols=lambda n: range(n//2)`
* Skip the second column, and move the first column to the end in the array that is returned:
      `usecols=lambda n: list(range(2, n)) + [0]`

I'm inclined to go with Eric's suggestion.  We might also want to accept a slice, to make it possible to express some of the common use-cases more concisely, but we could also live without it for a while, to first see if there is demand.

To make the use-case of skipping columns easier, we could provide a helper factory function, e.g. something like
```
def skip(skipcols):
    def skipper(n, skip=skipcols): 
        skip = [normalize_axis_index(c, n) for c in skip]
        usecols = [c for c in range(n) if c not in skip] 
        return usecols 
    return skipper
```
To skip the second column, one would use

    usecols=skip([1])

and to skip the second and last column,

    usecols=skip([1, -1])

etc.

What do folks think?  Any ideas for alternative APIs?

**Comment by eric-wieser** [comment]
> We might also want to accept a slice, to make it possible to express some of the common use-cases more concisely

Seems reasonable - so the logic would be something like:
```python
def normalize_usecols_to_list(usecols, n):
    try:
        return list(usecols)
    except TypeError:
        pass
    if isinstance(usecols, slice):
        return list(range(n)[usecols])
    elif callable(usecols):
        return list(usecols(n))
    else
        return [usecols]
```

**Comment by WarrenWeckesser** [comment]
@eric-wieser, yes, that's what I had in mind.

**Comment by WarrenWeckesser** [comment]
The above discussion also applies to `genfromtxt` (in case that wasn't obvious to anyone).

**Comment by WarrenWeckesser** [comment]
Here's what proposed API looks like in use:

```
In [1]: import numpy as np                                                                                                              

In [2]: np.savetxt('v.dat', np.vander([1, 2, 3, 4], N=6, increasing=True).T, fmt='%5d')                                                 

In [3]: !cat v.dat                                                                                                                      
    1     1     1     1
    1     2     3     4
    1     4     9    16
    1     8    27    64
    1    16    81   256
    1    32   243  1024

In [4]: np.loadtxt('v.dat', dtype=int, usecols=[1, -1])  # Nothing new here.                                                            
Out[4]: 
array([[   1,    1],
       [   2,    4],
       [   4,   16],
       [   8,   64],
       [  16,  256],
       [  32, 1024]])

In [5]: np.loadtxt('v.dat', dtype=int, usecols=np.s_[1::2])  # Use a slice                                                              
Out[5]: 
array([[   1,    1],
       [   2,    4],
       [   4,   16],
       [   8,   64],
       [  16,  256],
       [  32, 1024]])

In [6]: np.loadtxt('v.dat', dtype=int, usecols=lambda n: list(range(n // 2)))  # Function                                               
Out[6]: 
array([[ 1,  1],
       [ 1,  2],
       [ 1,  4],
       [ 1,  8],
       [ 1, 16],
       [ 1, 32]])

In [7]: np.loadtxt('v.dat', dtype=int, usecols=np.skip([1]))  # Helper for skipping columns                                             
Out[7]: 
array([[   1,    1,    1],
       [   1,    3,    4],
       [   1,    9,   16],
       [   1,   27,   64],
       [   1,   81,  256],
       [   1,  243, 1024]])

In [8]: np.loadtxt('v.dat', dtype=int, usecols=np.skip([1, -1]))  # Helper for skipping columns                                         
Out[8]: 
array([[  1,   1],
       [  1,   3],
       [  1,   9],
       [  1,  27],
       [  1,  81],
       [  1, 243]])
```

**Comment by eric-wieser** [comment]
Not really a fan of adding `np.skip` to the global namespace, but otherwise that proposal is sound. `np.lib.npio.skip` is a bit of a mouthful too.

**Comment by WarrenWeckesser** [comment]
@eric-wieser agreed.  I put it there to test, but the purpose of the functon is too specific to put in the top-level namespace, especially with such a generic name.

Is `numpy.lib.npyio` officially a public namespace?  It is easy enough to document and add examples that do

    from numpy.lib.npyio import skip

**Comment by WarrenWeckesser** [comment]
Pull request with the enhancement for loadtxt: https://github.com/numpy/numpy/pull/15995

**Comment by seberg** [comment]
Just to note, pandas allows a callable which gets the column names (and returns a boolean array).  So there is some precedence, although a bit different than this.
